"""
DESCRIPTION
"""
#__author__ = "Victoria Cepeda

#configfile: expand("config.json",

#include_prefix="https://"
#include_prefix + "rules"

#os.system("touch %s"%config['reads'][0])


ruleorder: kmer_mask > fastq2fasta > reference_recruitment > bowtie2_map > build_contigs > sam_to_bam > bam_sort > pilon_contigs

#code to skip initial steps if reference genomes provided 
if config['r1'][0] != "" and config['ref'] != "%s"%expand('{prefix}/{sample}.0.assembly.out/mc.refseq.fna',sample=config['sample'],prefix=config["prefix"])[0]:

     os.system("touch %s"%expand('{prefix}/{sample}.marker.match.1.fastq',prefix=config['prefix'],sample=config['sample'])[0])
     os.system("touch %s"%expand('{prefix}/{sample}.fasta',prefix=config['prefix'],sample=config['sample'])[0])
     os.system("mkdir -p %s"%expand('{prefix}/{sample}.0.assembly.out',prefix=config['prefix'],sample=config['sample'])[0])
     os.system("mkdir -p %s"%expand('{prefix}/{sample}.{iter}.assembly.out',prefix=config['prefix'],sample=config['sample'],iter=config['iter'])[0])
     os.system("touch %s"%expand('{prefix}/{sample}.0.assembly.out/mc.refseq.fna',prefix=config['prefix'],sample=config['sample'])[0])

rule all:
    input:expand('{prefix}/{sample}.{iter}.assembly.out/contigs.pilon.fasta',sample=config["sample"],prefix=config["prefix"],iter=config["iter"])
    benchmark:
       "%s/benchmarks/all/%s.txt"%(config['prefix'],config['sample'])

#expand('{reads}',reads=config['r1'])[0]
rule kmer_mask:
    input:
        r1=config['r1'],
        r2=config['r2']
    output:
        fastq1=expand('{prefix}/{sample}.marker.match.1.fastq',prefix=config['prefix'],sample=config['sample'])[0],
        fastq2=expand('{prefix}/{sample}.marker.match.2.fastq',prefix=config['prefix'],sample=config['sample'])[0]
    message: """---kmer-mask fastq"""
    benchmark:
       "%s/benchmarks/kmermask/%s.txt"%(config['prefix'],config['sample'])
    params:
        out=expand('{prefix}/{sample}.marker',prefix=config['prefix'],sample=config['sample']),
        len=str(config["length"]+3)
    threads:config["nthreads"]

    shell : "kmer-mask -ms 28 -mdb %s/refseq/kmer-mask_db/markers.mdb -1 {input.r1} -2 {input.r2} -clean 0.0 -match 0.01 -nomasking -t {threads} -l {params.len} -o {params.out}"%(config["mcdir"])


rule fastq2fasta:
    input: rules.kmer_mask.output.fastq1
    output:expand('{prefix}/{sample}.fasta',prefix=config['prefix'],sample=config['sample'])
    message: """---Converting fastq to fasta."""
    shell : "perl %s/bin/fq2fa.pl -i {input} -o {output}"%(config["mcdir"])


rule reference_recruitment:
    input:
        rules.fastq2fasta.output
    benchmark:
       "%s/benchmarks/reference_recruitment/%s.txt"%(config['prefix'],config['sample'])
    output:
        out =expand('{prefix}/{sample}.{iter}.assembly.out',prefix=config['prefix'],sample=config['sample'],iter=config['iter']),
	reffile =expand('{prefix}/{sample}.0.assembly.out/mc.refseq.fna',prefix=config['prefix'],sample=config['sample'])
    #fqfile =expand('{prefix}/{sample}.0.assembly.out/{sample}.fq',prefix=config['prefix'],sample=config['sample'])
    #'{prefix}/{sample}.{iter}.assembly.out/mc.refseq.fna'
    log:'%s/%s.%s.step1.log'%(config['prefix'],config['sample'],config['iter'])
    message: """---reference recruitment."""
    threads:config["nthreads"]

    shell:"mkdir -p {output.out}; %s/bin/pickrefseqs.pl {input} {output.out} {threads}  1> {log} 2>&1"%(config["mcdir"])


#reads=rules.reference_recruitment.output.fqfile
rule bowtie2_map:
    input:
       ref=rules.reference_recruitment.output.reffile,
       r1=config['r1'],
       r2=config['r2']
    benchmark:
       "%s/benchmarks/bowtie2_map/%s.txt"%(config['prefix'],config['sample'])
    output:
       index= '%s/%s.%s.assembly.out/%s.index'%(config['prefix'],config['sample'],config['iter'],config['sample']),
       pref='%s/%s.%s.assembly.out/%s.index'%(config['prefix'],config['sample'],config['iter'],config['sample']),
       sam='%s/%s.%s.assembly.out/%s.sam'%(config['prefix'],config['sample'],config['iter'],config['sample'])
    log: '%s/%s.%s.step2.log'%(config['prefix'],config['sample'],config['iter'])
    threads:config["nthreads"]
    message: """---Build index ."""
    shell:"bowtie2-build --threads {threads} -q %s {output.pref} 1>> {output.index} 2>&1;bowtie2 --end-to-end --very-sensitive --no-unal -p {threads} -x {output.pref} -q -1 {input.r1} -2 {input.r2} -S {output.sam} > {log} 2>&1"%(config['ref'])


#shell:"bowtie2-build --threads {threads} -q {input.ref} {output.pref} 1>> {output.index} 2>&1;bowtie2 --end-to-end -D 15 -R 2 -N 0 -L 31 -i S,1,1.15 -p {threads} -x {output.pref} -f {input.reads} -S {output.sam} > {log} 2>&1"


rule build_contigs:
    input:
        genome = '%s'%(config['ref']),
        sam=  rules.bowtie2_map.output.sam
    benchmark:
       "%s/benchmarks/build_contigs/%s.txt"%(config['prefix'],config['sample'])
    params:
        pickref="%s"%(config['pickref'])
    output:
        out='%s/%s.%s.assembly.out'%(config['prefix'],config['sample'],config['iter']),
        contigs='%s/%s.%s.assembly.out/contigs.fasta'%(config['prefix'],config['sample'],config['iter'])
    log:'%s/%s.%s.assembly.out/%s.assembly.log'%(config['prefix'],config['sample'],config['iter'],config['sample'])
    threads:1
    message: """---Build contigs ."""
    shell:"%s/bin/buildcontig -r {input.genome} -s {input.sam} -o {output.out} -c 2 -l 300 -n T -b F -u F -k {params.pickref}  1>> {log} 2>&1"%(config["mcdir"])

rule sam_to_bam:
    input:
        sam=  rules.bowtie2_map.output.sam
    output:
        bam = "%s.bam"%(rules.bowtie2_map.output.sam)
    log:'%s/%s.%s.assembly.out/%s.assembly.log'%(config['prefix'],config['sample'],config['iter'],config['sample'])
    threads:1
    message: """---Convert sam to bam ."""
    shell:"samtools view -bS {input.sam} -s {input.sam} -o {output.bam} 1>> {log} 2>&1"

rule bam_sort:
    input:
        bam = rules.sam_to_bam.output.bam 
    output:
        bam_sorted = "%s/%s.%s.assembly.out/sorted.bam"%(config['prefix'],config['sample'],config['iter'])
    log:'%s/%s.%s.assembly.out/%s.assembly.log'%(config['prefix'],config['sample'],config['iter'],config['sample'])
    threads:1
    message: """---Sort bam ."""
    shell: "samtools sort {input.bam} %s/%s.%s.assembly.out/sorted 1>> {log} 2>&1; samtools index {output.bam_sorted} 1>> {log} 2>&1"%(config['prefix'],config['sample'],config['iter'])


#samtools view -bS $samfile | samtools sort - $pilon_dir/$ref.sorted"
#$cmd = "samtools index $pilon_dir/$ref.sorted.bam";

rule pilon_contigs:
    input:
        contigs=rules.build_contigs.output.contigs,
        sam = rules.bam_sort.output.bam_sorted
    benchmark:
       "%s/benchmarks/pilon_contigs/%s.txt"%(config['prefix'],config['sample'])
    params:
        threads="%s"%(config['nthreads'])
    output:
        pilonctg='%s/%s.%s.assembly.out/contigs.pilon.fasta'%(config['prefix'],config['sample'],config['iter'])
    log:'%s/%s.%s.assembly.out/%s.pilon.log'%(config['prefix'],config['sample'],config['iter'],config['sample'])
    threads:1
    message: """---Pilon contigs ."""
    shell:"python %s/bin/fixhdr.py {input.contigs} ;java -Xmx16G -jar %s/bin/pilon-1.18.jar --threads {threads} --mindepth 10 --genome {input.contigs}.fna --frags {input.sam} --output %s/%s.%s.assembly.out/contigs.pilon --fix bases,breaks --vcf --changes 1>> {log} 2>&1"%(config["mcdir"],config["mcdir"],config['prefix'],config['sample'],config['iter'])

